docs/proposals/20250513-propogate-taints.md

```diff
@@ -0,0 +1,455 @@
+
+```golang
+type ControlPlaneClass struct {
+  // taints are the node taints that Cluster API will manage.
+  // ...
+  Taints []MachineTaint `json:"taints,omitempty"`
+
+  // Other fields...
+}
+
+type MachineDeploymentClass struct{
+  // taints are the node taints that Cluster API will manage.
+  // ...
+  Taints []MachineTaint `json:"taints,omitempty"`
+
+  // Other fields...
+}
+
+type MachinePoolClass struct{
+  // taints are the node taints that Cluster API will manage.
+  // ...
+  Taints []MachineTaint `json:"taints,omitempty"`
+
+  // Other fields...
+}
+```
+
+#### Proposed contract changes
+
+##### Changes to the bootstrap provider contract
+
+Document that `Initialize` taints are only added when a bootstrap provider supports the `node.cluster.x-k8s.io/uninitialized` taint.
+For details, see below paragraph "[Changes to the Machine, MachineSet, MachineDeployment and MachinePool resources via MachineSpec](#changes-to-the-machine-machineset-machinedeployment-and-machinepool-resources-via-machinespec)".
+
+##### Changes to the ControlPlane contract
+
+To support taints for control plane Machines, the contract for ControlPlanes should be extended.
+
+Similar to the extension for `ReadinessGates` at the `Machines` section, an additional optional  `FooControlPlaneMachineTemplate` rule should be added for supporting taints.
+
+Support and usage of that field would result in having the defined taints set on the Machines created by the ControlPlane provider and the Machine controller being responsible to behave as for worker Machines.
+
+#### Proposed controller changes
+
+##### Changes to the Machine and MachinePool controllers
+
+The Machine and MachinePool controllers as of today already manage the lifecycle of the taint `node.cluster.x-k8s.io/uninitialized`.
+
+These controllers should make use of that taint and add all `Initialize` taints at the same time as they remove it from a corresponding Node.
+This ensures that a node ever only gets these taints added once at the time of finishing initialization of a Node.
+
+If a Node re-registers itself and this includes having the `node.cluster.x-k8s.io/uninitialized`, the controllers also would again add the `Initialize` taints.
+
+Also all `Always` taints should get set for the first time, before or at the time of the removal of `node.cluster.x-k8s.io/uninitialized`.
+
+Also at this stage the controllers could check and (re-)add all taints of the `Always` type.
+
+Note: With this change we should also add the information to the "Contract rules for BootstrapConfig" documentation, that `Initialize` taints are only added when a bootstrap provider supports  the `node.cluster.x-k8s.io/uninitialized`.
+
+To keep track of which taints have been added by the controller, the controller also adds the annotation `cluster.x-k8s.io/taints-from-machine` to the Node.
+This follows the convention established by `cluster.x-k8s.io/labels-from-machine`.
```

Comment:
trying to create a hypothetical scenario (which might also be a problem for labels and other propagated fields),
if a set of taints are propagated to the node by a CAPI controller, but then a third party controller changes one of the taints. now we have set of taints that a tracked as owned by CAPI according to the annotation, but would that be the accurate state of the world, given the third party controller made a change too?

Reply:
As we are continuously writing the always taints it would be accurate.

We should probably block some more taints (probably entire domains like we do for labels and annotations)

Reply:
> node.kubernetes.io/* is one good candidate to block?

Seems reasonable to me. @chrischdi lets add it?

Reply:
I think it's correct to say that every CAPI Machine that is a control plane must have that label. Because that's how the label is defined. If someone doesn't do that then their implementation is buggy.

So accordingly having the CP label is a prereq for being allowed to set the CP taint.

Reply:
Added a contract PR: #12893

And updated the proposal

Reply:
Thx. All good from my side. I'll leave this conversation open in case someone else wants to comment further